### PR TITLE
fix: core fixes during SDK dev

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

- Normalizing thirdParty config as it may contain nulls while parsing from json.
- Making session APIs app specific
- Adding flags to fetch / revoke user session across all tenants

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core config (similar to the `access_token_signing_key_update_interval` config alias).
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/session/Session.java [hunk 1]
```diff
@@ -669,59 +671,144 @@ public static String[] revokeSessionUsingSessionHandles(Main main,
...
             }
 
-            if (tenantIdentifierWithStorage.getSessionStorage()
-                    .getSession(tenantIdentifierWithStorage, sessionHandle) == null) {
-                result[indexIntoResult] = sessionHandle;
-                indexIntoResult++;
+            sessionHandleMap.get(tenantId).add(sessionHandle);
+        }
+
+        int numberOfSessionsRevoked;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:697]:
```suggestion
```

File: src/main/java/io/supertokens/session/Session.java [hunk 2]
```diff
@@ -669,59 +671,144 @@ public static String[] revokeSessionUsingSessionHandles(Main main,
...
+        int numberOfSessionsRevoked = tenantIdentifierWithStorage.getSessionStorage()
+                .deleteSession(tenantIdentifierWithStorage, sessionHandles);
+
+        // most of the time we will enter the below if statement
+        if (numberOfSessionsRevoked == sessionHandles.length) {
+            return sessionHandles;
+        } else if (numberOfSessionsRevoked == 0) {
+            return new String[0];
+        } else {
+            String revokedSessionHandles[] = new String[numberOfSessionsRevoked];
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:746]: make it an array list and then return the to array of that at the end

File: src/main/java/io/supertokens/session/Session.java [hunk 3]
```diff
@@ -669,59 +671,144 @@ public static String[] revokeSessionUsingSessionHandles(Main main,
...
     @TestOnly
     public static String[] getAllNonExpiredSessionHandlesForUser(Main main, String userId)
             throws StorageQueryException {
         Storage storage = StorageLayer.getStorage(main);
-        return getAllNonExpiredSessionHandlesForUser(
-                new TenantIdentifierWithStorage(null, null, null, storage), userId);
+        return getAllNonExpiredSessionHandlesForUser(main,
+                new AppIdentifierWithStorage(null, null, storage), userId);
+    }
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:788]: add tests for non multi tenancy use (basically these funcitons being called when multi tenancy is not enabled)

File: src/main/java/io/supertokens/webserver/api/session/JWTDataAPI.java [hunk 4]
```diff
@@ -99,7 +104,7 @@ protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IO
     @Override
     @Deprecated
     protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
-        // API is tenant specific
+        // API is app specific but tenant id is derived from the session handle
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/session/JWTDataAPI.java:107]: make change in this API

File: src/main/java/io/supertokens/webserver/api/session/SessionRemoveAPI.java [hunk 5]
```diff
@@ -66,6 +66,11 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
             }
         }
 
+        Boolean revokeAcrossAllTenants = InputParser.parseBooleanOrThrowError(input, "revokeAcrossAllTenants", true);
+        if (revokeAcrossAllTenants == null) {
+            revokeAcrossAllTenants = true;
+        }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/session/SessionRemoveAPI.java:72]: if this is specified in the input, and the sessionHandles is specified in the input, throw a 400 error

File: src/test/java/io/supertokens/test/session/api/MultitenantAPITest.java [hunk 6]
```diff
@@ -424,4 +426,77 @@ public void testAccessTokensContainsTid() throws Exception {
             assertEquals(t2.getTenantId(), accessTokenInfo.payload.get("tId").getAsString());
         }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/session/api/MultitenantAPITest.java:427]: add tests where multi tenancy is not enabled and we are calling these functions.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/session/Session.java (hunk 1 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/Session.java:697]: Done

File: src/main/java/io/supertokens/session/Session.java (hunk 2 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/Session.java:746]: Done

File: src/main/java/io/supertokens/webserver/api/session/JWTDataAPI.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/session/JWTDataAPI.java:107]: Done

File: src/main/java/io/supertokens/webserver/api/session/SessionRemoveAPI.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/session/SessionRemoveAPI.java:72]: Done

File: src/main/java/io/supertokens/session/Session.java [hunk 7]
```diff
@@ -26,12 +26,12 @@
 import io.supertokens.exceptions.TryRefreshTokenException;
 import io.supertokens.exceptions.UnauthorisedException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:27]: patch version change

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Pull request merged**